pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -472,9 +472,12 @@ func (e *EndpointController) syncService(key string) error {
 
 	createEndpoints := len(currentEndpoints.ResourceVersion) == 0
 
+	// Compare the sorted subsets and labels
+	// Remove the HeadlessService label from the endpoints if it exists, as this won't be set on the service itself
+	// and will cause a false negative in this diff check
 	if !createEndpoints &&
-		apiequality.Semantic.DeepEqual(currentEndpoints.Subsets, subsets) &&
-		apiequality.Semantic.DeepEqual(currentEndpoints.Labels, service.Labels) {
+		apiequality.Semantic.DeepEqual(endpoints.RepackSubsets(currentEndpoints.Subsets), subsets) &&
```

Comment:
This shouldn't be needed - endpoints controller is creating endpoints with already sorted subsets.

So as long, as the endpoint has been created by endpoints controller, this isn't needed. If this wasn't then that sounds like two things fighting with each other, so it's also a problem.

So I think this isn't needed.

Reply:
Theoretically I'd agree, but the test clusters were we experienced it proof otherwise. We don't have any external actors modifying endpoints, yet the comparison here causes excessive updates to all endpoints. Not sure why @damemi went with `RepackSubsets` since we initially discussed only the necessity to sort them. I'll catch up with Mike when he's back online. This is especially visible on bigger clusters with +10k, but also reproducible on smaller ones.

Reply:
> We don't have any external actors modifying endpoints, yet the comparison here causes excessive updates to all endpoints.

Hmm - so I think that means we have some other bug somewhere else. So instead of this mitigation, I would say that we should try understand why is this happening...

Reply:
Are we sure it's sorting endpoints before creating/updating them? And are they sorted when they return from the `endpointsLister` call above? (https://github.com/kubernetes/kubernetes/pull/94112/files#diff-a1a9c0efe93384ed9a010e65e8ad4604L459)

Reply:
@robscott @freehan to confirm, but it seems we're sorting them before creating/updated.

[and obviously endpointsLister is returning whatever we committed to etcd earlier via creations/updates]

Reply:
@soltysh @damemi - thanks for the data

Hmm - this looks super strange...
I looked into the code a bit more, and we are sorting ports by hash when repacking in both places:
https://github.com/kubernetes/kubernetes/blob/14a11060a0775ed609f0810898ebdbe737c59441/pkg/api/v1/endpoints/util.go#L197
https://github.com/kubernetes/kubernetes/blob/14a11060a0775ed609f0810898ebdbe737c59441/pkg/api//endpoints/util.go#L196
[ideally, we should try to combine these two functions, as there is high-probability of divergence here - @robscott ]

So that looks fine.
Neither of our client machinery (informers, listers, etc.) should NOT modify the objects, these are supposed to only cache them and not touch anything.
Deserialization code looks fine, so the deserialized object should actually preserve the order of port.

> Having said that, I think it's most reliable (and efficient too) if our controllers do not rely on a specific order, as long as data is correct. @wojtek-t wdyt?

I'm slightly afraid that this would mask some real issue somewhere in the system. I would much rather prefer to track it down first. @liggitt - WDYT?

Reply:
> Hmm - so I think that means we have some other bug somewhere else. So instead of this mitigation, I would say that we should try understand why is this happening...

I agree with this. We need to root cause the issue rather than compensating for it.

Reply:
Thanks for identifying this issue @damemi! It seems like we really need to understand what is causing the change in Endpoints order. Have you been able to tell if any Endpoints updates are happening that cause the change in port ordering? This does feel like a lower level issue than the controller itself, so would be good to dig a bit further to understand what's happening. You've mentioned that this is most noticeable in larger clusters, is there a way to consistently reproduce this behavior at smaller scales?

Reply:
I spoke shortly with Jordan about it on slack, I'll try have another look into it, focusing on what might be causing this mutation.

Reply:
I'm looking at things that could explain a difference between the server canonicalized order and the controller-computed order:

* Are there any non-nil empty lists in the object when the controller packs subsets? If so, that could change the computed hash in https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/api/endpoints/util.go#L198, then those non-nil empty lists would be nil when the server observed them (because protobuf serialization does not distinguish between nil and non-nil empty lists), and the server could repack in a different order
* I see a lot of uses of `sort.Sort` by RepackSubsets, which does not guarantee stable sort order for items whose sort keys are identical (e.g. endpoint addresses which share an IP and targetRef.UID) ... do identical items like that exist in the cases you are seeing?
* Is the controller sending an object with empty fields that will have defaulting applied to them (e.g. https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/apis/core/v1/defaults.go#L321-L331)? That seems like it could influence the hash and order

Reply:
I'm able to reproduce it using the integration test framework https://github.com/kubernetes/kubernetes/pull/94277
The problem seems to be with the endpoint strategy
```
TEST endpoint strategy [{[] [{1.1.1.1  0xc00566ba50 0xc000556380}] [{port-1337 1337 TCP <nil>} {port-1338 1338 TCP <nil>}]}]
TEST endpoint strategy after [{[] [{1.1.1.1  0xc00566ba50 0xc000556380}] [{port-1338 1338 TCP <nil>} {port-1337 1337 TCP <nil>}]}]
```

seems @wojtek-t was right https://github.com/kubernetes/kubernetes/pull/94112/files#r475439845

Reply:
if I skip `Canonicalize()` then it works, 
```
I0827 13:36:28.154446  192183 endpoints_controller.go:475] TEST endpoints are equal?
 current subsets: [{[] [{1.1.1.1  0xc004ec7150 &ObjectReference{Kind:Pod,Namespace:test-endpoints-updates,Name:test-pod,UID:3082f47b-4a7b-4b69-bae8-56cee3fdf0ba,APIVersion:,ResourceVersion:124,FieldPath:,}}] [{port-1337 1337 TCP <nil>} {port-1338 1338 TCP <nil>}]}]
 subsets [{[] [{1.1.1.1  0xc004bfa1d8 &ObjectReference{Kind:Pod,Namespace:test-endpoints-updates,Name:test-pod,UID:3082f47b-4a7b-4b69-bae8-56cee3fdf0ba,APIVersion:,ResourceVersion:124,FieldPath:,}}] [{port-1337 1337 TCP <nil>} {port-1338 1338 TCP <nil>}]}]
I0827 13:36:28.154559  192183 endpoints_controller.go:480] endpoints are equal for test-endpoints-updates/foo, skipping update
```

Reply:
> I'm able to reproduce it using the integration test framework #94277

excellent... that will make tracking down the issue much easier

> if I skip `Canonicalize()` then it works

can you dump the JSON serialization of the endpoints object in the following places:
* before the controller calls repack
* after the controller calls repack
* before Canonicalize calls repack
* after Canonicalize calls repack

Reply:
I also noticed few additional things:
* we've had a v1 and internal version of RepackSubsets since bff3cb4df68d699285285c09c1c46be69ef1465a, which is problematic if they drift ... we should verify they still match and put fuzzing tests in place to make sure the equivalent v1/internal versions pack identically
* anything using `hashObject` is vulnerable to skew between the controller and API server when a new field (like AppProtocol) is added to one of the types... the newer component will hash based on the new field with a nil value, and the older component will omit that from the hash
  * any changes to RepackSubsets/SortSubsets need to be done *extremely* carefully, and in a way that does not trigger this issue for all endpoints when in a skew state. Our lack of skew testing makes us especially vulnerable here.

Reply:
looks like there's a diff between the v1 and internal implementations:

```
import (
	"reflect"
	"testing"

	"github.com/davecgh/go-spew/spew"
	"github.com/google/go-cmp/cmp"
	v1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/types"
	v1util "k8s.io/kubernetes/pkg/api/v1/endpoints"
	api "k8s.io/kubernetes/pkg/apis/core"
	apiv1 "k8s.io/kubernetes/pkg/apis/core/v1"
	"k8s.io/utils/pointer"
)

func TestSortSubsetsCompatibility(t *testing.T) {
	// Start with v1 endpoints
	v1Endpoints := &v1.Endpoints{
		Subsets: []v1.EndpointSubset{{
			NotReadyAddresses: []v1.EndpointAddress{{
				IP:       "1.1.1.1",
				NodeName: pointer.StringPtr("fakenode"),
				TargetRef: &v1.ObjectReference{
					Kind:            "Pod",
					Namespace:       "test-endpoints-updates",
					Name:            "test-pod",
					UID:             "277ecc82-f0a9-4d57-92da-5966466d1afc",
					ResourceVersion: "126828",
				},
			}},
			Ports: []v1.EndpointPort{
				{Name: "port-1337", Port: 1337, Protocol: "TCP"},
				{Name: "port-1338", Port: 1338, Protocol: "TCP"},
			},
		}},
	}

	// Pack with v1 util, matching what the controller will do
	v1Endpoints.Subsets = v1util.RepackSubsets(v1Endpoints.Subsets)

	// Convert to internal
	internalEndpoints := &api.Endpoints{}
	if err := apiv1.Convert_v1_Endpoints_To_core_Endpoints(v1Endpoints.DeepCopy(), internalEndpoints, nil); err != nil {
		t.Fatal(err)
	}

	// Save a copy to compare
	beforeInternal := internalEndpoints.DeepCopy()

	// Pack with internal util, matching with the API server will do
	internalEndpoints.Subsets = RepackSubsets(internalEndpoints.Subsets)
	// Should be no diff
	if diff := cmp.Diff(beforeInternal.Subsets, internalEndpoints.Subsets); len(diff) > 0 {
		t.Error(diff)
	}
}
```

```sh
go test ./pkg/api/endpoints -run TestSortSubsetsCompatibility
--- FAIL: TestSortSubsetsCompatibility (0.00s)
    util_test.go:71:   []core.EndpointSubset{
          	{
          		Addresses:         nil,
          		NotReadyAddresses: []core.EndpointAddress{{IP: "1.1.1.1", NodeName: &"fakenode", TargetRef: &core.ObjectReference{Kind: "Pod", Namespace: "test-endpoints-updates", Name: "test-pod", UID: "277ecc82-f0a9-4d57-92da-5966466d1afc", ResourceVersion: "126828"}}},
          		Ports: []core.EndpointPort{
          			{
        - 				Name:        "port-1337",
        + 				Name:        "port-1338",
        - 				Port:        1337,
        + 				Port:        1338,
          				Protocol:    "TCP",
          				AppProtocol: nil,
          			},
          			{
        - 				Name:        "port-1338",
        + 				Name:        "port-1337",
        - 				Port:        1338,
        + 				Port:        1337,
          				Protocol:    "TCP",
          				AppProtocol: nil,
          			},
          		},
          	},
          }
        
FAIL
FAIL	k8s.io/kubernetes/pkg/api/endpoints	0.284s
```

Reply:
One issue is with object hash:

The v1 implementation is hashing this string:
> `([]endpoints.addrReady)[{addr:(*v1.EndpointAddress){IP:(string)1.1.1.1 Hostname:(string) NodeName:(*string)fakenode TargetRef:(*v1.ObjectReference){Kind:(string)Pod Namespace:(string)test-endpoints-updates Name:(string)test-pod UID:(types.UID)277ecc82-f0a9-4d57-92da-5966
466d1afc APIVersion:(string) ResourceVersion:(string)126828 FieldPath:(string)}} ready:(bool)false}]`

The internal implementation is hashing this string:
> `([]endpoints.addrReady)[{addr:(*core.EndpointAddress){IP:(string)1.1.1.1 Hostname:(string) NodeName:(*string)fakenode TargetRef:(*core.ObjectReference){Kind:(string)Pod Namespace:(string)test-endpoints-updates Name:(string)test-pod UID:(types.UID)277ecc82-f0a9-4d57-92da-
5966466d1afc APIVersion:(string) ResourceVersion:(string)126828 FieldPath:(string)}} ready:(bool)false}]`

Note the different type names:
* `v1.EndpointAddress` vs `core.EndpointAddress`
* `v1.ObjectReference` vs `core.ObjectReference`

For that reason, and because of the version skew issue noted in https://github.com/kubernetes/kubernetes/pull/94112/files#r478397569, using hashes of the go struct seems really problematic here.

Reply:
So, to summarize, I see the following issues that need to be fixed if we continue to normalize order both in the server and controller and expect them to match perfectly in order to settle:

* hashing go structs differs from internal/external struct type names
* hashing go structs differs when there's skew between the API server version and controller version with new fields
* hashing go structs differs from nil empty slices and non-nil empty slices with protobuf collapsing the difference between those
* not using sort.Stable doesn't guarantee the same sort order for values whose sort keys are identical
* server-side defaulting applied before server normalization can affect hash values and make them differ from the controller-computed order
* independent implementations of normalization for v1 (used by controller) and internal (used by API server) can drift

Reply:
> we've had a v1 (used by the controller) and internal (used by the API server) version of RepackSubsets since bff3cb4, which is problematic if they drift .

Yeah - I also mentioned that above in:  https://github.com/kubernetes/kubernetes/pull/94112#discussion_r475439845


BTW - nice list above, thanks Jordan!

Reply:
Sine the biggest issue is around hasing objects in `RepackSubsets` I took two different approaches and here are the results:
1. Remove types in the string representation coming from spew library in [DeepHashObject](https://github.com/kubernetes/kubernetes/blob/72350a3cf7b1b021eb31d72ab98013ebe8455fc4/pkg/util/hash/hash.go#L28).
2. Convert `v1.EndpointSubset` to `core.EndpointSubset`, invoke internal version of `RepackSubsets` and convert back to v1.

Benchmark results of those are:
1. 17737 ns/op
2. 13305 ns/op

as a comparison, the current implementation is running roughly 13222 ns/op. (Those are results from my env.)

The first is  ~30% slower, rather naive approach and fragile to changes. Second one on the other hand gives us similar performance to what we currently have, and most importantly removes code duplication. We could probably provide the both versions and ensure the conversions are hidden from users. 
For second solution I was also considering having a 3rd `RepackSubset` internal representation, but I'm not sure that's something we'd like to pursuit. 

@liggitt @wojtek-t @thockin @robscott opinions?

Reply:
Three of the issues could be addressed by converting/using a single Repack method (and using `sort.Stable` everywhere):

* [x] internal/external struct type names differ, causing hash differences
* [x] independent implementations of normalization for v1 (used by controller) and internal (used by API server) can drift
* [x] not using sort.Stable doesn't guarantee the same sort order for values whose sort keys are identical

These issues would still remain:
* [ ] version skew between the API server and controller with new fields in the structs causes differences
* [ ] nil empty slices and non-nil empty slices (caused by protobuf collapsing the difference between those) causes differences
* [ ] server-side defaulting applied before server normalization can affect hash values and make them differ from the controller-computed order

If we were going to convert/use a single repack method, I would lean toward having the server convert to v1.Endpoints and use the method intended for use with the versioned object.

I'm actually wondering if it makes more sense to drop the server-side canonicalization if re-ordering is the only change being made (I could see keeping it in deduplication cases, which would make it obvious the client was not sending what it thought was a repacked subset).

Reply:
> I'm actually wondering if it makes more sense to drop the server-side canonicalization if re-ordering is the only change being made (I could see keeping it in deduplication cases, which would make it obvious the client was not sending what it thought was a repacked subset).

I was suggesting dropping it completely, but what you suggest makes more sense to me.
So basically, if there aren't duplicates, leave as is, if there are duplicates, do repacking.

Reply:
I can definitely support the idea of dropping the server-side canonicalization, but I'm not sure about the history behind it. Maybe @thockin or @robscott might provide them? 

In general endpoints controller is quite frequently going through all of them, so even if somebody posts unordered Endpoints, it will be corrected in a timely fashion.

Reply:
I've talked about this with @freehan and neither of us can think of any reason endpoints would need to be consistently sorted. I can not find anything inside kube-proxy that requires this kind of specific ordering. It feels surprising and inconsistent that resources would be modified in this way before they were saved. I can potentially understand the desire to leave deduplication in as part of the strategy, but full repacking at this point seems both unnecessary and unexpected.

Reply:
I thought the sorting had to do with making it easier to find the relative complement of the old and new sets, in order to be able to update _iptables_ and _ipvs_ rules.

Reply:
> Theoretically I'd agree, but the test clusters were we experienced it proof otherwise. We don't have any external actors modifying endpoints, yet the comparison here causes excessive updates to all endpoints. Not sure why @damemi went with `RepackSubsets` since we initially discussed only the necessity to sort them. I'll catch up with Mike when he's back online. This is especially visible on bigger clusters with +10k, but also reproducible on smaller ones. 

We also saw the same thing. Please see the _Additional Info_ section in https://github.com/kubernetes/kubernetes/issues/94339.

I spent just about three weeks debugging this issue (the aforementioned section describes why it took so long), and it turns out we were working on close to the same problem the entire time :)

Reply:
@thockin it seems that you've added the repacking of the endpoints in the apiserver in #5939  ?to deal with the migration to multi-port endpoints? 
based on the comments in this thread it seems that there is no reason now to keep doing it. Can you confirm if is ok to _drop the server-side canonicalization_?
If affirmative, @damemi @soltysh  I went ahead and have a commit https://github.com/kubernetes/kubernetes/pull/94277/commits/fba1e3f0b65d9395e06a1bce171c7602a114d264  dropping the canonicalization in the apiserver and  a test case covering, feel free to take and add it to this PR

Reply:
I see no harm in the endpoint canonicalization in-and-of-itself. It may still serve a purpose. The issue, IMO, is the conflict with the ServerSideApply described in #94339. Without SSA enabled, this is a non-issue -- the canonicalization works as intended and additional updates are no-ops. It is only once SSA was enabled by default that this became a problem.

Which raises a much larger issue -- should clients be expected to know and reproduce any server-side storage strategies and their data mutation logic to avoid superfluous `MODIFIED` events? My vote would be no, I think this is an unreasonable expectation to place upon users. Since SSA is likely not going anywhere, and storage strategies can and do make sense, shouldn't we try to make the two work together?

Reply:
> I see no harm in the endpoint canonicalization in-and-of-itself. It may still serve a purpose.

Github is eliding the comment, but there are six issues identified with this specific server-side normalization - https://github.com/kubernetes/kubernetes/pull/94112#discussion_r478422983

> The issue, IMO, is the conflict with the ServerSideApply described in #94339.

That is an additional issue/interaction which needs to be resolved, but is orthogonal.

Reply:
> there are six issues identified with this specific server-side normalization

ack

> That is an additional issue/interaction which needs to be resolved, but is orthogonal.

ack